### PR TITLE
feat(llma): add tools tab for LLM Analytics

### PR DESCRIPTION
**Pull request by andrewm4894**
## Summary

- Extract tool call names from `$ai_output_choices` at ingestion time into `$ai_tools_called` (comma-separated) and `$ai_tool_call_count` (integer) properties
- Handles 7 output formats: OpenAI Chat, OpenAI unwrapped, OpenAI Responses API, Normalized SDK, Anthropic, Python repr (Agents SDK), and wrapped full API response (`{"choices":[...]}`)
- Hardened with fast keyword pre-checks, 500KB size limit, tool name sanitization, and 100-tool cap
- Add a new Tools tab (behind `LLM_ANALYTICS_TOOLS_TAB` feature flag) showing per-tool aggregation with solo % indicating how often a tool is called alone vs in a batch

[Demo Video](https://drive.google.com/file/d/1pLSIzs8TCSc0QfRYXf4Q_IqbIkNoiYYQ/view?usp=sharing)

### Ingestion (Node.js)
- `nodejs/src/ingestion/ai/tools/extract-tool-calls.ts` — Core extraction for 7 formats, sanitization (trim, comma→underscore, 200-char truncate), 100-tool cap
- `nodejs/src/ingestion/ai/tools/index.ts` — Pipeline step with fast keyword pre-check (`tool_call`, `tool_use`, `function_call`, `"function"`), 500KB size limit, new metric labels (`skipped_no_indicators`, `skipped_too_large`)
- `nodejs/src/ingestion/ai/tools/extract-tool-calls.test.ts` — 66 parameterized tests
- `nodejs/src/ingestion/ai/process-ai-event.ts` — Wired as pipeline step 5
- `nodejs/src/ingestion/ai/metrics.ts` — `aiToolCallExtractionCounter` Prometheus counter

### Frontend
- `products/llm_analytics/frontend/tabs/llmAnalyticsToolsLogic.ts` — Kea logic (mirrors errors tab pattern)
- `products/llm_analytics/frontend/LLMAnalyticsTools.tsx` — DataTable with tool name links to filtered generations, solo % renderer
- `products/llm_analytics/frontend/LLMAnalyticsScene.tsx` — Tab entry behind feature flag
- `products/llm_analytics/frontend/llmAnalyticsSharedLogic.ts` — Added `tools` tab ID and URL routing
- `products/llm_analytics/manifest.tsx` — Route and URL
- `frontend/src/lib/constants.tsx` — `LLM_ANALYTICS_TOOLS_TAB` feature flag

### Backend
- `products/llm_analytics/backend/queries/tools.sql` — HogQL query using `arrayJoin(splitByChar(',', ...))` with columns: tool, total_calls, solo_pct, traces, users, sessions, days_seen, first_seen, last_seen
- `posthog/taxonomy/taxonomy.py` — Property definitions for `$ai_tools_called` and `$ai_tool_call_count`

### Storage format
Comma-separated string (not JSON array). HogQL's `JSONExtractRaw` + regex stripping leaves escaped quotes that break `JSONExtractArrayRaw`. Comma-separated works cleanly with `splitByChar`.

## Test plan

- [x] 66 unit tests covering all 7 formats, sanitization, caps, pre-checks, wrapped responses, malformed data, and edge cases
- [ ] Enable `LLM_ANALYTICS_TOOLS_TAB` feature flag and verify the tab renders at `/llm-analytics/tools`
- [ ] Send test `$ai_generation` events with tool calls and verify extraction works for all formats
- [ ] Verify clicking a tool name navigates to generations filtered by that tool
- [ ] Verify solo % column shows correct values (100% for always-solo tools, lower for batch tools)

### CODE REVIEW

File: products/llm_analytics/frontend/LLMAnalyticsTools.tsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                                <Tooltip title={`View generations calling ${toolString}`}>
+                                    <Link
+                                        to={
+                                            combineUrl(urls.llmAnalyticsGenerations(), {
+                                                ...searchParams,
+                                                filters: [
+                                                    {
+                                                        type: PropertyFilterType.Event,
+                                                        key: '$ai_tools_called',
+                                                        operator: PropertyOperator.IContains,
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/frontend/LLMAnalyticsTools.tsx]: Fixed in c1aadc98 — switched from `IContains` to `Regex` with `(^|,)toolname(,|$)` so it matches exact comma-delimited tokens. Since we use comma-separated storage, this gives precise matching without substring false positives.

File: nodejs/src/ingestion/ai/tools/index.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    // Only process $ai_generation events
+    if (event.event !== '$ai_generation') {
+        return event
+    }
+
+    try {
+        // Respect user-provided values
+        if (event.properties['$ai_tools_called'] !== undefined && event.properties['$ai_tools_called'] !== null) {
+            return event
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply nodejs/src/ingestion/ai/tools/index.ts:58]: Good catch — fixed in c1aadc98. Now when `$ai_tools_called` is user-provided but `$ai_tool_call_count` is missing, we derive the count by splitting the comma-separated string. This ensures these events aren't excluded by the `$ai_tool_call_count > 0` filter in the tools query.

File: products/llm_analytics/frontend/LLMAnalyticsTools.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                                                        value: toolString,
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/frontend/LLMAnalyticsTools.tsx]: Same fix as above — switched to `Regex` with `(^|,)toolname(,|$)` in c1aadc98.

File: nodejs/src/ingestion/ai/tools/index.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        const toolNames = extractToolCallNames(parsed, rawString)
+
+        if (toolNames.length === 0) {
+            return event
+        }
+
+        event.properties['$ai_tools_called'] = toolNames.join(',')
+        event.properties['$ai_tool_call_count'] = toolNames.length
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply nodejs/src/ingestion/ai/tools/index.ts:102]: Already handled — `sanitizeToolName()` in `extract-tool-calls.ts` replaces commas with underscores before storage, so tool names like `get,weather` become `get_weather`. This was added in bf9b0dc as part of the hardening work. No commas can survive into the stored value.

File: products/llm_analytics/backend/queries/tools.sql [hunk 5]
```diff
@@ -0,0 +1,23 @@
+/*
+-- Tool call extraction is done at ingestion time.
+-- The $ai_tools_called property contains comma-separated tool names.
+-- See: nodejs/src/ingestion/ai/tools/extract-tool-calls.ts
+*/
+
+SELECT
+    arrayJoin(splitByChar(',', assumeNotNull(properties.$ai_tools_called))) as tool,
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/backend/queries/tools.sql:12]: Good catch — added `AND tool \!= ''` after the `arrayJoin` to guard against empty tool names from edge cases. The `$ai_tools_called \!= ''` guard is redundant since `$ai_tool_call_count > 0` already covers that, but the post-split check is the one that matters.

File: products/llm_analytics/frontend/LLMAnalyticsTools.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                                                ...searchParams,
+                                                filters: [
+                                                    {
+                                                        type: PropertyFilterType.Event,
+                                                        key: '$ai_tools_called',
+                                                        operator: PropertyOperator.Regex,
+                                                        value: `(^|,)${toolString}(,|$)`,
+                                                    },
+                                                ],
+                                            }).url
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply products/llm_analytics/frontend/LLMAnalyticsTools.tsx:67]: Good catch — fixed in 5b95e2fc8f using the existing `escapeRegex` utility from `lib/actionUtils`.

File: nodejs/src/ingestion/ai/tools/index.ts
Comment (carlos-marchal-ph) [inline nodejs/src/ingestion/ai/tools/index.ts:1]: There is some inconsistency to when we increment the metrics for this with `aiToolCallExtractionCounter`. We are not doing it when:
- We extract the count only from a user provided `$ai_tools_called`
- We attempt extraction but there is not `$ai_output_choices` field
- There are no tools calls in the output choices

File: posthog/taxonomy/taxonomy.py
Comment (carlos-marchal-ph) [inline posthog/taxonomy/taxonomy.py:1]: After updating this, file it's necessary to run `pnpm taxonomy:build`, otherwise the frontend doesn't get updated:

Image
vs
Image

**Review by carlos-marchal-ph — APPROVED** [review]
Looks good! I was a bit confused at first by the "Solo" column, had to read the description on hover to figure it out. Not saying it's not useful, but since it's the only percentage in the table I intuitively assumed that it would be the % of conversations in which it was called.

Which wouldn't be the most useful column either if the customer has different types of conversations or sessions with different tool calls expected in each, but still my head went there first for some reason.

File: nodejs/src/ingestion/ai/tools/index.ts
Comment (andrewm4894) [reply nodejs/src/ingestion/ai/tools/index.ts:1]: Fixed in c1a2b265 — every code path now increments `aiToolCallExtractionCounter` with a descriptive status: `user_provided`, `no_output_choices`, `no_tools_found`, `skipped_too_large`, `skipped_no_indicators`, `extracted`, and `error`.

File: posthog/taxonomy/taxonomy.py
Comment (andrewm4894) [reply posthog/taxonomy/taxonomy.py:1]: Fixed in c1a2b265 — ran `pnpm taxonomy:build` to regenerate the frontend JSON. The `$ai_tools_called` and `$ai_tool_call_count` properties now show proper labels and descriptions in the UI.
